Travis ci

shubhamshuklaer opened the pull request:
~~**This still needs some work, So don't merge yet.**~~**This can be merged**
**This contains 2 commits from #305 . Once that is merged I will rebase it off of master.**
Fixes #302 

For a working version check
https://github.com/shubhamshuklaer/ghi/tree/travis-ci
https://travis-ci.org/shubhamshuklaer/ghi

For merging check
https://github.com/shubhamshuklaer/ghi/tree/travis-ci#enable-travis-ci-in-fork

I am almost done, but I wanted to get suggestions from others. If anyone have any idea on how to improve it, please comment. Right now there are only the very basic tests. I won't be able to add any more tests myself(as time ran out), but if anyone wants to contribute they can open a pull request on `travis-ci` branch on https://github.com/shubhamshuklaer/ghi or wait for this to get merged and then open a pull request on this repo.

Because of the nature of the software, the tests overlap with each other. For eg. for testing `ghi edit` you first need to create an issue, maybe a milestone too and then edit that issue. As a result editing issue also tests issue creation and milestone creation. But I have also kept separate tests for issue creation and milestone creation, so that they can be tested separately.

shubhamshuklaer committed:
Take title and description on cmdline for milestone just like open

Now `ghi milestone sample_title -m sample_description` will create a
milestone with title as sample_title and description as
sample_description. Fixes #303

shubhamshuklaer committed:
Take title and description on cmdline for edit just like open

shubhamshuklaer commented on README.md:
```diff
@@ -1,5 +1,7 @@
 # ghi
 
+[![Build Status](https://travis-ci.org/shubhamshuklaer/ghi.svg?branch=travis-ci)](https://travis-ci.org/shubhamshuklaer/ghi)
```
This needs to be changed after merge.

davidCarlos commented:
Nice job @shubhamshuklaer, i will take a look and try to contribute some how. I think that we could squash some of those commits, cause will be difficult to review all this code with so much commits.

shubhamshuklaer commented:
@davidCarlos thanks. You are right. I will reduce the no of commits. Right now I am working on improving local testing. Since the tester will use a fake github account for testing and a genuine account for using ghi, this will cause clash. I was thinking of suggesting the testers to use a different user(computer user) for testing and then that user will have different `~/.gitconfig` and different env vars values for `GITHUB_USER` and `GITHUB_PASSWORD`. Do you have a better way??

shubhamshuklaer commented:
One more thing I will fix is the indentation style. I doesn't match the rest of the project. I will fix it before squashing commits.

shubhamshuklaer commented:
@davidCarlos can also evaluate whether the information in readme is enough to setup for local testing as well as setting up travis-ci for fork.

shubhamshuklaer commented:
One problem is that the secure(We marked display these in build log to false) env varriables `GITHUB_USER` and `GITHUB_PASSWORD` are not provided by travis while building pull requests from fork. The explanation is given here https://docs.travis-ci.com/user/pull-requests#Security-Restrictions-when-testing-Pull-Requests .  Hence tests cannot run for pull requests from forks(It will run when the pull request is from the same repo).

Any suggestions for solving it??
One way is to not make it secure. But then it will be public and anyone will be able to see it.

shubhamshuklaer commented:
One more thing that needs to be fixed is deleting the tokens once done with the tests(We can add a flag if the user don't want to delete it). Since travis spins up different containers on each run(of test suit), each run gets a new token and a lot of tokens are accumulated in github. Its best to delete them(For security reasons as well as management).
Right now the token is generated before running the first test. And its not deleted cause we don't know which is the last test. 
The way to fix it will be to generate the token before each test and to delete it after each test.

shubhamshuklaer commented:
### Fixed the following things
* Corrected the Indentation style of tests(Now following project's style)
* Improved local testing(Tokens are not added nor used from ~/.gitconfig while testing) so you can both test with fake account and use ghi normally with original account simultaneously.
* Deleting tokens after running tests.
 
### TODO
* squash commits

shubhamshuklaer committed:
Add .editorconfig to maintain project's coding style

shubhamshuklaer committed:
Add --just_print_token option to ghi config --auth

Helpful while doing local testing. We don't want to mess with the user's
original ~/.gitconfig. User will want to use a different github account
for testing(Don't want to mess up the original account).

shubhamshuklaer committed:
Setup communication with github

shubhamshuklaer committed:
Setup mock data and its helper functions

shubhamshuklaer committed:
Setup repo and token creation/deletion

shubhamshuklaer committed:
Test opening issue

shubhamshuklaer committed:
Test comment creation/ammend/deletion

shubhamshuklaer committed:
Test milestone creation

shubhamshuklaer committed:
Test [un-]assigning user to the issue

shubhamshuklaer committed:
Test editing issue

shubhamshuklaer committed:
Test adding/deleting/replacing labels from issue

shubhamshuklaer committed:
Test closing issue

shubhamshuklaer committed:
Test showing issue

shubhamshuklaer committed:
Test listing issue

shubhamshuklaer committed:
A helper script for cleaning up(deleting all repos)

shubhamshuklaer committed:
Update Readme

shubhamshuklaer committed:
Setup framework for travis-ci and local testing

stephencelis commented:
This is great! And what an undertaking! If it's complete (less the README link), I wonder if @AlexChesters has time to look it over and connect it to the PR mechanism here (to ensure future PRs don't break the build).

shubhamshuklaer commented on .editorconfig:
```diff
@@ -0,0 +1,12 @@
+trim_trailing_whitespace = true
+
+[*.rb]
+indent_style = tab
```
Looks like the project uses space instead of tabs. I will fix this.

shubhamshuklaer committed:
Convert tabs to spaces. The project uses spaces

shubhamshuklaer commented:
@stephencelis 
Thanks. Its complete from my side(less the README link).
 About PR mechanism, (as mentioned in a previous comment) since the test environment contains 2 secure env vars namely `GITHUB_USER` and `GITHUB_PASSWORD` , travis doesn't provide them to pull request from forks(It considers them insecure).
 So the test won't work like in other repos where they automatically run tests as soon as the PR is made. 
However we can run tests manually before merging by pulling the changes to a temporary local branch and then pushing it to github(then the travis will run on that branch) or run the tests locally using `rake test:one_by_one GITHUB_USER='xyz' GITHUB_PASSWORD='abc'`.

shubhamshuklaer commented:
The reason for keeping the username and password secure(Not available publically) even though its a fake account.
https://github.com/shubhamshuklaer/ghi/issues/3

> 
Even if the user is fake, it still linked to the creator and if we make username and password public someone can misuse it. For eg
    1. uploading bad things to that account
    2. run a script which will continuously keep making api request to it, forcing github people to suspend the account.
    3. changing the password(Just for causing problems).
    4. Using that account to post bad things everywhere(Other's repos) on github.
>
Even if we encrypt the username and password, it will be de-crypted before running tests, so anyone can write a test which will print it and run that test locally.

stephencelis commented:
@shubhamshuklaer Been thinking on this. One possibility to avoid the need to run with an actual user/password would be to use the VCR gem to record the API requests:

https://github.com/vcr/vcr

shubhamshuklaer commented:
@stephencelis That is an awesome idea. But I don't know when I will be able to do it. 

If someone else is interested in doing it, that would be awesome.

olleolleolle commented on README.md:
```diff
@@ -69,3 +71,47 @@ FAQs can be found in the [wiki](https://github.com/stephencelis/ghi/wiki/FAQ)
 
 ![Example](images/example.png)
+
+## Testing Guidlines
```
typo: `Guidlines` => `Guidelines`

olleolleolle commented on README.md:
```diff
@@ -69,3 +71,47 @@ FAQs can be found in the [wiki](https://github.com/stephencelis/ghi/wiki/FAQ)
+* Before running tests `GITHUB_USER` and `GITHUB_PASSWORD` environment
+variables must be exported. It is best to use a fake account as a bug can mess
+up your original account. You can either export these 2 environment variables 
+through `~/.bashrc`(As ghi only uses these while generating its token, so after
```
Space missing before `(As`

olleolleolle commented on README.md:
```diff
@@ -69,3 +71,47 @@ FAQs can be found in the [wiki](https://github.com/stephencelis/ghi/wiki/FAQ)
+variables must be exported. It is best to use a fake account as a bug can mess
+up your original account. You can either export these 2 environment variables 
+through `~/.bashrc`(As ghi only uses these while generating its token, so after
+you generate the token for your original account(for regular use), fake account
```
Space missing before `(for`

olleolleolle commented on README.md:
```diff
@@ -69,3 +71,47 @@ FAQs can be found in the [wiki](https://github.com/stephencelis/ghi/wiki/FAQ)
+* Open a Travis CI account and activate travis-ci for the fork
+* Create a fake github account for testing. The username, password and token
+will be available to the tests and if by mistake the test prints it, it will be
+available in public log. So its best to create a fake account and use a
```
typo `its` => `it's`

olleolleolle commented on README.md:
```diff
@@ -69,3 +71,47 @@ FAQs can be found in the [wiki](https://github.com/stephencelis/ghi/wiki/FAQ)
+* At Travis-CI, on the settings page for the fork, add environment variables
+`GITHUB_USER` and `GITHUB_PASSWORD`. Ensure that the "Display value in build
+log" is set to false. It is possible to add these in ".travis.yml", but don't
+as all forks as well as original repo will be using different accounts(We cannot 
```
Space missing before `(We`

olleolleolle commented on lib/ghi/commands/config.rb:
```diff
@@ -10,6 +10,9 @@ def options
           opts.on '--local', 'set for local repo only' do
             assigns[:local] = true
           end
+          opts.on '--just_print_token', "Just print the token don't add it to ~/.gitconfig(Useful while testing)" do
```
Space missing before `(Useful`

olleolleolle commented on test/assign_test.rb:
```diff
@@ -0,0 +1,42 @@
+require "test/unit"
+require "helper"
+require "pp"
```
`pp` no longer used in this file.

olleolleolle commented on test/close_test.rb:
```diff
@@ -0,0 +1,30 @@
+require "test/unit"
+require "helper"
+require "pp"
```
`pp` no longer used in this file.

olleolleolle commented on test/comment_test.rb:
```diff
@@ -0,0 +1,45 @@
+require "test/unit"
+require "helper"
+require "pp"
```
`pp` not used.

olleolleolle commented on test/delete_all_repos.rb:
```diff
@@ -0,0 +1,30 @@
+require "helper"
+require "json"
+require "pp"
```
`pp` not used.

olleolleolle commented:
@shubhamshuklaer I wanted to say thanks ✨  by adding a slew of specific notes in the code. Thanks!

shubhamshuklaer commented:
@olleolleolle Thanks a lot for the review. I am a bit out of touch with this PR right now. Will fix this when incorporating stephencelis' idea of using vcr gem.
Thanks
